### ISSUE TITLE
Artificial Kuramoto Oscillatory Neurons

### ISSUE THREAD

**Issue by zhuoxiao-chen** [zhuoxiao-chen/please_read_paper_every_day#65]
<img width="1123" alt="image" src="https://github.com/user-attachments/assets/47cfcd24-0529-49e2-ae14-75846648dc9c">

### Paper Link

[arxiv] https://arxiv.org/abs/2410.13821

[OpenReview] https://openreview.net/forum?id=nwDRD4AMoN

### Code Link

https://github.com/autonomousvision/akorn

## Brief Introduction

* This paper introduces Artificial Kuramoto Oscillatory Neurons (AKOrN) based on Kuramoto model, which can be can be combined with layer architectures such as fully connected layers, convolutions, and attention mechanisms.
* This idea provides performance improvements across a wide spectrum of tasks such as unsupervised object discovery, adversarial robustness, calibrated uncertainty quantification, and reasoning.

### From neuroscience to Kuramoto model.
- We ask ourselves again if there are lessons from neuroscience that can be used as design principles for artificial neural nets.
- It was recognized early on that **neurons interact** via lateral connections. In fact, neighboring neurons tend to cluster their activities and clusters tend to compete to explain the input. 
- This process is made possible by synchronization: like fireflies in the night, neurons tend to synchronize their activities with their neighbors’, which leads to the compression of their representations.
- In this paper, we will use an N-dimensional generalization of the famous Kuramoto model (Kuramoto, 1984) to model this synchronization.

### Kuramoto Model 

- We denote **oscillators** by $\mathbf{X} = \{\mathbf{x}_i\}_{i=1}^C$, where each $\mathbf{x}_i$ is a vector on a **hypersphere**: $\mathbf{x}_i \in \mathbb{R}^N$, $\|\mathbf{x}_i\|_2 = 1$. $N$ is each single oscillator dimension called **rotating dimensions** and $C$ is the **number of oscillators**. 
- The **oscillator index** $i$ may have multiple dimensions: if the input is an image, for example, each oscillator is represented by $\mathbf{x}_{c,h,w}$ with $c, h, w$ indicating channel, height, and width positions, respectively.

The differential equation of our vector-valued Kuramoto model is written as follows:
$$
\dot{\mathbf{x}}_i = \mathbf{\Omega}_i \mathbf{x}_i + \mathrm{Proj}_{\mathbf{x}_i} \left(\mathbf{c}_i + \sum_j J_{ij} \mathbf{x}_j \right), \quad \text{where} \quad \mathrm{Proj}_{\mathbf{x}_i} (\mathbf{y}_i) = \mathbf{y}_i - \langle \mathbf{y}_i, \mathbf{x}_i \rangle \mathbf{x}_i
$$

- $\mathbf{\Omega}_i$ is an $N \times N$ anti-symmetric matrix and $\mathbf{\Omega}_i \mathbf{x}_i$ is called the **natural frequency** term that determines each oscillator’s **own** rotation frequency and angle. 
- The second term governs **interactions** between oscillators, where $\mathrm{Proj}_{\mathbf{x}_i}$ is an operator that projects an input vector onto the **tangent space of the sphere** at $\mathbf{x}_i$. 
- $J_{ij} \in \mathbb{R}$ represents the connections between oscillators: if $J_{ij} > 0$, the $i$-th and $j$-th oscillators tend to align, and if $J_{ij} < 0$, they tend to oppose each other.

<img width="1300" alt="image" src="https://github.com/user-attachments/assets/9ba79bdb-2338-48ff-864e-d1a2cc7b1c8d">

- $\mathbf{C} = \{\mathbf{c}_i\}_{i=1}^C, \mathbf{c}_i \in \mathbb{R}^N$ is a data-dependent variable, which is computed from the observational input or the activations of the previous layer. 

<img width="1302" alt="image" src="https://github.com/user-attachments/assets/061ce0bd-7de3-4701-a3e1-25ca53010c8a">

The Kuramoto model is Lyapunov if we assume certain symmetric properties in $J_{ij}$ and $\mathbf{\Omega}_i$ 
(Aoyagi, 1995; Wang \& Roychowdhury, 2017). For example, if $\mathbf{J}$ is symmetric and different oscillators share the same natural frequencies: $J_{ij} = J_{ji}^\top$, $\mathbf{\Omega}_i = \mathbf{0}$, and $\mathbf{\Omega} \mathbf{c}_i = \mathbf{0}$, each update is guaranteed to minimize the following energy:
$$
E = -\sum_{i,j} \mathbf{x}_i^\top J_{ij} \mathbf{x}_j - \sum_i \mathbf{c}_i^\top \mathbf{x}_i
$$

<img width="1291" alt="image" src="https://github.com/user-attachments/assets/f0c05e70-2fe7-45ef-9cd5-d92a5834a34e">

**Finding**: The oscillator state is initially disordered, but gradually exhibits collective behavior, eventually becoming a spatially propagating **wavy pattern**.

> Our model has the advantage that it naturally incorporates spatiotemporal representations **in the form of traveling waves**, for which there is ample evidence in the neuroscientific literature. Muller et al. (2016) finds that **oscillatory patterns** in the thalamocortical network during sleep are organized into circular wave-like patterns, which could give an account of how memories are consolidated in the brain. Davis et al. (2020) suggest that spontaneous traveling waves in the visual cortex modulate synaptic activities and thus act as a gating mechanism in the brain. **In the generalized Kuramoto model, traveling waves naturally emerge as neighboring oscillators start to synchronize.**

> Another advantage of using dynamical neurons is that they can perform a form of **reasoning**. Kuramoto oscillators have been successfully used to solve combinatorial optimization tasks such as kSAT problems. Many authors have argued that the modern architectures based on, e.g., transformers lack this intrinsic capability of “neuro-symbolic reasoning. We show that AKOrN can successfully solve Sudoku puzzles, illustrating this capability. Additionally, AKOrN relates to models in quantum physics and active matter (see appendix B).

## Networks with Kuramoto Oscillator

<img width="1305" alt="image" src="https://github.com/user-attachments/assets/fa3235a4-4d15-4038-8f6d-25ecb2d02c5f">

- We utilize the artificial Kuramoto oscillator neurons (AKOrN) as a basic unit of information processing in neural networks (Fig. 2). 
- First, we transform an observation with a **relatively simple function** to create the initial conditional stimuli $\mathbf{C}^{(0)}$. 
- Next, $\mathbf{X}^{(0)}$ is initialized by either $\mathbf{C}^{(0)}$, a fixed learned embedding, random vectors, or a mixture of these initialization schemes. 
- The block is composed of two modules: the Kuramoto layer and the readout module, which together process the pair $\{\mathbf{X}, \mathbf{C}\}$. 
- The Kuramoto layer updates $\mathbf{X}$ with the conditional stimuli $\mathbf{C}$, and the readout layer extracts features from the final oscillatory states to create new conditional stimuli. 
We denote the number of layers by $L$, and the $l$-th layer’s output by $\{\mathbf{X}^{(l)}, \mathbf{C}^{(l)}\}$.

<img width="1143" alt="image" src="https://github.com/user-attachments/assets/b7579102-f790-48a3-96df-ab85ce33d01d">

> For each layer, update oscillators T times then pass to next layer.

<img width="1302" alt="image" src="https://github.com/user-attachments/assets/99d080fc-278a-4227-82eb-de859f7ad36d">

> At time T, compute norm of weighted X as C to capture phase invariant patterns. 

> Specifically, each m is the norm of **z**, and **z** is the sum of weighted **x**. 

### Implement AKOrN in Comvolution and Attention Layer

<img width="1187" alt="image" src="https://github.com/user-attachments/assets/70d34c87-b2c7-4cd2-a7fd-fef6e6e47b65">

# Experiments

## Setting

- We observe that both the readout module and conditional stimuli $C$ are essential for stable training, especially when $N = 2$. 
- We also see that AKOrN with $N = 2$ exhibits a strong regularity, which acts positively on robustness performance while having negative effects on unsupervised object discovery and the Sudoku-solving experiments. We show results of AKOrN with $N = 4$ in those experiments. 
- We do not observe improvement by increasing $N$ above 4. 

## Unsupervised Object Discovery

- Unsupervised object discovery is the task of finding objects in an image without supervision.
- We test on 4 synthetic benchmark datasets (Tetrominoes, dSprites, CLEVR, CLEVRTex), one synthetic dataset created by us (Shapes), and 2 real image benchmark datasets (PascalVOC, COCO2017).
- As baselines, we train models that are similar to ResNet (He et al., 2016) and ViT (Dosovitskiy et al., 2021), but **iterate** the convolution or self-attention layers multiple times with shared parameters. This allows us to evaluate the impact of our proposed, Kuramoto-based iterative updates. We denote these baselines as Iterative Convolution (ItrConv) and Iterative Self-Attention (ItrSA), respectively. 

<img width="1171" alt="image" src="https://github.com/user-attachments/assets/d5be1d05-737d-42ab-b03e-dc91693be425">

- In AKOrN, $C$ is initialized by the patched features of the images, while each $x_i$ is initialized by random oscillators sampled from the **uniform** distribution on the sphere. We train the AKOrN model with the self-supervised SimCLR (Chen et al., 2020) objective.
- We train each model from scratch on the four synthetic datasets.
- For the two real image datasets, we directly evaluate that ImageNet pretrained model on both datasets without fine-tuning.
- When evaluating, we apply clustering to the final block’s output features (In AKOrN, it is $C^{(L)}$ ). 
- We use agglomeration clustering with average linkage, which we found to outperform K-means for both the baseline models and AKOrN.
- We evaluate the clustering results by FG-ARI and MBO.
- FG-ARI (foreground adjusted rand index) measures the similarity between the ground truth masks and the computed clusters, only for foreground objects. 
- MBO (Mean-Best-Overlap) first assigns each cluster to the highest overlapping ground truth mask and then computes the average intersection-over-union (IoU) of all pairs.

<img width="895" alt="image" src="https://github.com/user-attachments/assets/7b1a5109-9475-44c0-a7dd-a164d9acffae">

- Fig. 3 shows that AKOrNs improve the object discovery performance over their non-Kuramoto counterparts in every dataset. Interestingly, we observe that convolution is less effective than attention.
- In Fig. 4, we see that the Kuramoto models’ clusters are well aligned with the individual objects, while clusters of the ItrSA model often span across objects and background, and are sensitive to the texture of the background and the specular highlight on the floor

<img width="1170" alt="image" src="https://github.com/user-attachments/assets/d378db89-9473-4253-9b9d-d745b8c76110">

- Fig. 5 shows AKOrN binds object features on natural images much better than DINO.

<img width="500" alt="image" src="https://github.com/user-attachments/assets/89ddb11b-4003-47f0-926b-dd9775149ecb">

- For PascalVOC and COCO2017, we show instance-level MBO (MBO i ) and class-level (MBO c ) segmentation results.
- The proposed AKOrN model outperforms existing SSL models including DINO, MoCoV3, and MAE on both datasets, showing that it learns more object-bound features than conventional transformer-based models.

<img width="487" alt="image" src="https://github.com/user-attachments/assets/43962ba4-ae15-46b8-822f-11b9ebb8de3a">
<img width="492" alt="image" src="https://github.com/user-attachments/assets/9dba5d9c-2300-4b3b-bc04-04942f21d03e">

## SUDOKU 数独

<img width="1046" alt="image" src="https://github.com/user-attachments/assets/b4b88e3f-9bbe-4118-aa5b-ed0642674cfd">

<img width="500" alt="image" src="https://github.com/user-attachments/assets/dfb10117-cd8d-49f6-a804-0e0d1359c8fb">

## Robustness and Calibration

- Robustness to adversarial attacks and uncertainty quantificitation performance on the network with the and CIFAR10 with common corruptions.
- (AKOrN^conv) a has three convolutional Kuramoto blocks.
- (AKOrN^mix ) replaces the last block with an attentive Kuramoto block.
- We use AutoAttack (Caron et al., 2021) to evaluate the model’s adversarial robustness.

<img width="987" alt="image" src="https://github.com/user-attachments/assets/acaa0e6b-6ca3-4360-b20c-e928833ad7a1">

- **Tab. 4**: AKOrNs are resilient against gradient-based attacks
- **Fig. 8**: AKOrNs are robust to strong random noise (Fig. 8) 
- **Fig. 9**: good uncertainty estimation

## More Ablattion and Component Analysis

<img width="1079" alt="image" src="https://github.com/user-attachments/assets/f7e2e163-f68c-49ee-a01c-477ec973cd90">

<img width="1107" alt="image" src="https://github.com/user-attachments/assets/2a7bb903-7485-4a2f-a2d6-ade801d7ec35">

<img width="1045" alt="image" src="https://github.com/user-attachments/assets/f7e7c0c6-b931-480c-a15b-235572778837">

<img width="1097" alt="image" src="https://github.com/user-attachments/assets/d7ca38d8-40d7-40fe-8852-78664d9f95a6">
